xla/hlo/transforms/collectives/collective_quantizer_test.cc

```diff
@@ -449,5 +564,105 @@ TEST_F(CollectiveQuantizerTest, DequantizeAllGatherUnary) {
+      module->entry_computation()->root_instruction()->operand(0)->operand(0);
+  EXPECT_THAT(all_gather->shape().element_type(), F8E4M3FN);
+}
+
+TEST_F(CollectiveQuantizerTest, DequantizeAllToAllPartialReplication) {
+  absl::string_view hlo_string = R"(
+  HloModule module
+  max {
+    a = f32[] parameter(0)
+    b = f32[] parameter(1)
+    ROOT max = f32[] maximum(a, b)
+  }
+
+  ENTRY entry {
+    param = f8e4m3fn[8,32,8,128] parameter(0)
+    convert = bf16[8,32,8,128] convert(param)
+    scale = bf16[1] parameter(1), sharding={devices=[8]<=[8]}
+    scalar_scale = bf16[] reshape(scale)
+    all_reduced_scale = bf16[] all-reduce(scalar_scale), to_apply=max, replica_groups={{0,1,2,3},{4,5,6,7}}, channel_id=1, use_global_device_ids=true
+    scale_bcast = bf16[8,32,8,128] broadcast(all_reduced_scale), dimensions={}
+    multiply = bf16[8,32,8,128] multiply(convert, scale_bcast)
+    ROOT all-to-all = bf16[8,32,8,128] all-to-all(multiply), dimensions={1}, replica_groups={{0,1,2,3},{4,5,6,7}}, channel_id=2
+    }
+  )";
+
+  HloModuleConfig config;
+  config.set_num_partitions(8);
+  TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<HloModule> module,
+                          ParseAndReturnVerifiedModule(hlo_string, config));
+  TF_ASSERT_OK_AND_ASSIGN(bool changed, RunCollectiveQuantizer(module.get()));
+  EXPECT_TRUE(changed);
+  EXPECT_THAT(module->entry_computation()->root_instruction(),
+              op::Multiply(op::Convert(op::AllToAll(op::Parameter())),
+                           op::Broadcast()));
+  const HloInstruction* all_gather =
+      module->entry_computation()->root_instruction()->operand(0)->operand(0);
+  EXPECT_THAT(all_gather->shape().element_type(), F8E4M3FN);
+}
+
+TEST_F(CollectiveQuantizerTest,
+       DequantizeAllGatherPartialReplicationGroupMismatch) {
+  absl::string_view hlo_string = R"(
+  HloModule module
+  max {
+    a = f32[] parameter(0)
+    b = f32[] parameter(1)
+    ROOT max = f32[] maximum(a, b)
+  }
+
+  ENTRY entry {
+    param = f8e4m3fn[8,4,8,128] parameter(0)
+    convert = bf16[8,4,8,128] convert(param)
+    scale = bf16[1] parameter(1), sharding={devices=[8]<=[8]}
+    scalar_scale = bf16[] reshape(scale)
+    all_reduced_scale = bf16[] all-reduce(scalar_scale), to_apply=max, replica_groups={{0,1,2,3},{4,5,6,7}}, channel_id=1, use_global_device_ids=true
+    scale_bcast = bf16[8,4,8,128] broadcast(all_reduced_scale), dimensions={}
+    multiply = bf16[8,4,8,128] multiply(convert, scale_bcast)
+    ROOT all-gather = bf16[8,32,8,128] all-gather(multiply), dimensions={1}, replica_groups={{0,1,2,3,4,5,6,7}}, channel_id=2, use_global_device_ids=true
+    }
+  )";
+
+  HloModuleConfig config;
+  config.set_num_partitions(8);
+  TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<HloModule> module,
+                          ParseAndReturnVerifiedModule(hlo_string, config));
+  TF_ASSERT_OK_AND_ASSIGN(bool changed, RunCollectiveQuantizer(module.get()));
+  EXPECT_FALSE(changed);
+}
+
```

Comment:
It is worth testing the case where the scale is allreduced in another computation.